org.esbtools.auth.common/src/main/java/org/esbtools/auth/servlet/CertEnvironmentVerificationFilter.java

```diff
@@ -0,0 +1,75 @@
+package org.esbtools.auth.servlet;
+
+import java.io.IOException;
+import java.security.cert.X509Certificate;
+
+import javax.naming.NamingException;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
+
+import org.esbtools.auth.util.Environment;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CertEnvironmentVerificationFilter implements Filter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(CertEnvironmentVerificationFilter.class);
+
+  private final Environment env;
+
+  public CertEnvironmentVerificationFilter(String environment) {
+    env = new Environment(environment);
+
+    LOGGER.info("Cert Environment: " + ((environment == null) ? "Not Set" : environment));
+  }
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException {
+    //Do Nothing!
+  }
+
+  @Override
+  public void destroy() {
+    //Do Nothing!
+  }
+  
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+    LOGGER.debug("Attempting Environment Cert verification");
+    X509Certificate certChain[] = (X509Certificate[]) request.getAttribute("javax.servlet.request.X509Certificate");
+
+    if ((null != certChain) && (certChain.length > 0)) {
+      LOGGER.debug("Verifying environment on cert");
+      String dn = certChain[0].getSubjectDN().getName();
+      try {
+        env.validate(dn);
+      }
+      catch (NamingException e) {
+        unsuccessfulAuthentication(request, response, e);
+        return; //end the chain
+      }
+    }
+    else {
+      LOGGER.debug("Cert not found. Skipping Environment Cert verification.");
+    }
+
+    chain.doFilter(request, response);
+  }
+
+  protected void unsuccessfulAuthentication(ServletRequest request,
+          ServletResponse response, NamingException failed) throws IOException, ServletException {
+    if (response instanceof HttpServletResponse) {
+       ((HttpServletResponse) response).setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+    }
+    
+    response.setContentType("text/html");
```

Comment:
Should this maybe be JSON instead of HTML so it is more easily parsable by clients?

Reply:
I thought about that as well. The reason I used html is to be consistent with undertow, which uses html in all of its filter level responses.

My solution is a little convoluted here, but in the spring module I overwrote this class to throw an exception instead, which is caught by yet another filter. I am working on a hopefully better solution, but it might need to be a follow up to this PR.

Reply:
I hear you on consistency, I always hated how Undertow (and prior JBoss versions with tomcat) handled this, as it is painful for the clients who are expecting a JSON response

Reply:
Or I should say that @bvulaj is working on some changes that could simplify the CertEnvironmentVerificationFilter. So maybe this could happen in a follow up PR?

Reply:
This is not an essential change for this PR, as we have been living with this behavior for over three years now from JBoss, and don't get that many complaints.  Good to know @bvulaj is working on more changes to the CertEnvironmentVerificationFilter, let's pick this discussion back up when those are complete.